// swa/lib/meta.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
/**
 * @file meta 信息
 * @author ielgnaw(wuji0223@gmail.com)
 */

exports.default = {
    /**
     * 提示问题
     *
     * @type {Object}
     */
    PROMPTS: {
        name: {
            type: 'input',
            message: 'Project name'
        },
        description: {
            type: 'input',
            message: 'Project description',
            default: 'Project description'
        },
        author: {
            type: 'input',
            message: 'Author'
        },
        redis: {
            type: 'confirm',
            message: 'Use Redis'
        },
        auth: {
            type: 'list',
            message: 'Auth',
            choices: ['UUAP', 'Passport', 'none'],
            filter: function filter(val) {
                return val.toLowerCase();
            }
        }
        // separator: {
        //     message: 'Database'
        // },
    },

    /**
     * 依赖配置
     *
     * @type {Object}
     */
    ALL_DEPENDENCIES: {
        save: ['mysql'],
        saveDev: ['chai']
    }
};
module.exports = exports['default'];
