// swa/lib/command/init.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; /**
                                                                                                                                                                                                                                                                   * @file init 命令
                                                                                                                                                                                                                                                                   * @author ielgnaw(wuji0223@gmail.com)
                                                                                                                                                                                                                                                                   */

var _path = require('path');

var _fs = require('fs');

var _inquirer = require('inquirer');

var _inquirer2 = _interopRequireDefault(_inquirer);

var _rimraf = require('rimraf');

var _rimraf2 = _interopRequireDefault(_rimraf);

var _gitUser = require('../git-user');

var _gitUser2 = _interopRequireDefault(_gitUser);

var _meta = require('../meta');

var _meta2 = _interopRequireDefault(_meta);

var _generate = require('../generate');

var _generate2 = _interopRequireDefault(_generate);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * inquirer.prompt then function
 *
 * @param {string} projectName 项目名称
 * @param {boolean} inCurrent 是否在当前目录直接创建项目
 */
var inquirerPromptDone = function inquirerPromptDone(projectName, inCurrent) {
    var curPrompts = _extends({}, _meta2.default.PROMPTS);
    (0, _generate.setDefault)(curPrompts, 'name', projectName);
    (0, _generate.setDefault)(curPrompts, 'author', (0, _gitUser2.default)());
    (0, _generate2.default)(curPrompts, projectName, inCurrent);
};

exports.default = {
    command: ['init [directory]'],
    describe: 'Initialize project in the current directory or specified directory',
    builder: {},
    handler: function handler(argv) {
        var directory = argv.directory;
        if (!directory || directory === '.') {
            _inquirer2.default.prompt([{
                type: 'confirm',
                message: !(0, _fs.readdirSync)('.', 'utf8').length ? 'Generate project in current directory' : 'The current directory is not empty, do you want to overwrite',
                name: 'ok'
            }]).then(function (answers) {
                if (!answers.ok) {
                    return;
                }
                inquirerPromptDone((0, _path.basename)(process.cwd()), true);
            });
        } else {
            if ((0, _fs.existsSync)(directory) && (0, _fs.statSync)(directory).isDirectory()) {
                _inquirer2.default.prompt([{
                    type: 'confirm',
                    message: 'Target directory exists. Continue?',
                    name: 'ok'
                }]).then(function (answers) {
                    if (!answers.ok) {
                        return;
                    }
                    _rimraf2.default.sync((0, _path.join)('.', directory));
                    inquirerPromptDone(directory);
                });
                return;
            } else {
                inquirerPromptDone(directory);
            }
        }
    }
};
module.exports = exports['default'];
